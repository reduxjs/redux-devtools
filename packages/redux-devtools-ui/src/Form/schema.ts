import { JSONSchema7 } from 'json-schema';

export const schema: JSONSchema7 = {
  title: 'Example form',
  description: 'A simple form example.',
  type: 'object',
  required: ['name'],
  properties: {
    name: {
      type: 'string',
      title: 'Full name',
    },
    age: {
      type: 'integer',
      title: 'Age',
    },
    bio: {
      type: 'string',
      title: 'Bio',
    },
    password: {
      type: 'string',
      title: 'Password',
      minLength: 3,
    },
    multipleChoicesList: {
      type: 'array',
      title: 'A multiple choices list',
      items: {
        type: 'string',
        enum: ['foo', 'bar', 'fuzz'],
      },
      uniqueItems: true,
    },
    numberEnum: {
      type: 'number',
      title: 'Number enum',
      enum: [1, 2, 3],
    },
    numberEnumRadio: {
      type: 'number',
      title: 'Number enum',
      enum: [1, 2, 3],
    },
    integerRange: {
      title: 'Integer range',
      type: 'integer',
      minimum: 42,
      maximum: 100,
    },
  },
};

export const uiSchema = {
  name: {
    'ui:autofocus': true,
  },
  age: {
    'ui:widget': 'updown',
  },
  bio: {
    'ui:widget': 'textarea',
  },
  password: {
    'ui:widget': 'password',
    'ui:help': 'Hint: Make it strong!',
  },
  date: {
    'ui:widget': 'alt-datetime',
  },
  multipleChoicesList: {
    'ui:widget': 'checkboxes',
  },
  numberEnumRadio: {
    'ui:widget': 'radio',
    'ui:options': {
      inline: true,
    },
  },
  integerRange: {
    'ui:widget': 'range',
  },
};

export const formData = {
  name: 'Chuck Norris',
  age: 75,
  bio: 'Roundhouse kicking asses since 1940',
  password: 'noneed',
  integerRange: 52,
};
