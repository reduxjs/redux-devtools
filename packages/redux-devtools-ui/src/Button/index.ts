export { default } from './Button.js';
