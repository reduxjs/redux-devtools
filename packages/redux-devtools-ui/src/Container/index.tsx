import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { useTheme, ThemeData } from '../utils/theme';
import { MainContainerWrapper, ContainerWrapper } from './styles';
import { Theme } from '../themes/default';

interface ContainerFromThemeDataProps {
  children?: React.ReactNode;
  themeData: ThemeData;
  className?: string;
}

const ContainerFromThemeData: React.FunctionComponent<
  ContainerFromThemeDataProps
> = ({ themeData, className, children }) => {
  const theme = useTheme(themeData);
  return (
    <ThemeProvider theme={theme}>
      <MainContainerWrapper className={className}>
        {children}
      </MainContainerWrapper>
    </ThemeProvider>
  );
};

interface Props {
  children?: React.ReactNode;
  themeData?: ThemeData;
  theme?: Theme;
  className?: string;
}

const Container: React.FunctionComponent<Props> = ({
  themeData,
  className,
  theme,
  children,
}) => {
  if (!themeData) {
    return (
      <ContainerWrapper className={className} theme={theme}>
        {children}
      </ContainerWrapper>
    );
  }

  return (
    <ContainerFromThemeData themeData={themeData} className={className}>
      {children}
    </ContainerFromThemeData>
  );
};

Container.propTypes = {
  children: PropTypes.node,
  themeData: PropTypes.any,
  theme: PropTypes.any,
  className: PropTypes.string,
};

export default Container;
