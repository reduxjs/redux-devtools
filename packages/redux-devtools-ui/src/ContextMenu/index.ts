export { default } from './ContextMenu.js';
