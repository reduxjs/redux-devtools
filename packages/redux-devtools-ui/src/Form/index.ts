export { default } from './Form.js';
