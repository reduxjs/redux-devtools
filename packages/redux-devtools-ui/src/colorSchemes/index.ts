export { default } from './default';
export { default as github } from './github';
export { default as atomOneDark } from './atom-one-dark';
export { default as dracula } from './dracula';
export { default as iRBlack } from './ir-black';
export { default as macintosh } from './macintosh';
export { default as materia } from './materia';
export { default as oceanicNext } from './oceanic-next';
export { default as phD } from './phd';
export { default as pico } from './pico';
export { default as solarFlare } from './solar-flare';
export { default as spacemacs } from './spacemacs';
export { default as unikitty } from './unikitty';
export { default as woodland } from './woodland';
