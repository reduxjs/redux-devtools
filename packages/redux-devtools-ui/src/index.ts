export { default as Container } from './Container';
export { default as Button } from './Button';
export { default as ContextMenu } from './ContextMenu';
export { default as Dialog } from './Dialog';
export { default as Editor } from './Editor';
export { default as Form } from './Form';
export { default as Select } from './Select';
export { default as Slider } from './Slider';
export { default as Tabs, Tab } from './Tabs';
export { default as SegmentedControl } from './SegmentedControl';
export { default as Notification } from './Notification';
export * from './Toolbar';

import color from './utils/color';
export const effects = { color };
export { default as createStyledComponent } from './utils/createStyledComponent';
export { Theme, ThemeFromProvider, Scheme } from './utils/theme';
