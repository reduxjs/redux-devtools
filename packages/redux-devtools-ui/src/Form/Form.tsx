import React, { PureComponent, Component } from 'react';
import PropTypes from 'prop-types';
import JSONSchemaForm, { FormProps } from '@rjsf/core';
import { Base16Theme } from 'base16';
import createStyledComponent from '../utils/createStyledComponent';
import styles from './styles';
import Button from '../Button';
import customWidgets from './widgets';

const FormContainer = createStyledComponent(styles, JSONSchemaForm);

export interface Props<T> extends FormProps<T> {
  children?: React.ReactNode;
  submitText?: string;
  primaryButton?: boolean;
  noSubmit?: boolean;
  theme?: Base16Theme;
}

/**
 * Wrapper around [`react-jsonschema-form`](https://github.com/rjsf-team/react-jsonschema-form) with custom widgets.
 */
export default class Form<T> extends (PureComponent || Component)<Props<T>> {
  render() {
    const { widgets, children, submitText, primaryButton, noSubmit, ...rest } =
      this.props;
    return (
      <FormContainer
        {...(rest as Props<unknown>)}
        widgets={{ ...customWidgets, ...widgets }}
      >
        {noSubmit ? (
          <noscript />
        ) : (
          children || (
            <Button
              size="big"
              primary={primaryButton}
              theme={rest.theme}
              type="submit"
            >
              {submitText || 'Submit'}
            </Button>
          )
        )}
      </FormContainer>
    );
  }

  static propTypes = {
    children: PropTypes.any,
    submitText: PropTypes.string,
    primaryButton: PropTypes.bool,
    noSubmit: PropTypes.bool,
    schema: PropTypes.object.isRequired,
    uiSchema: PropTypes.object,
    formData: PropTypes.any,
    widgets: PropTypes.objectOf(
      PropTypes.oneOfType([PropTypes.func, PropTypes.object])
    ),
  };
}
