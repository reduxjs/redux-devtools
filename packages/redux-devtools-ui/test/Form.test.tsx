import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Form } from '../src';
import { schema, uiSchema, formData } from '../src/Form/schema';

describe('Form', function () {
  let random: () => number;

  beforeAll(() => {
    random = Math.random;
    Math.random = jest.fn(() => 0.25546350798039463);
  });

  afterAll(() => {
    Math.random = random;
    console.log(Math.random());
  });

  it('renders correctly', () => {
    const { container } = render(
      <Form formData={formData} schema={schema} uiSchema={uiSchema} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with primary button', () => {
    const { container } = render(
      <Form
        primaryButton
        submitText="Custom button"
        formData={formData}
        schema={schema}
        uiSchema={uiSchema}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with no button', () => {
    const { container } = render(
      <Form formData={formData} schema={schema} uiSchema={uiSchema} noSubmit />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should handle the submit event', () => {
    const onSubmit = jest.fn();
    render(
      <Form
        formData={formData}
        schema={schema}
        uiSchema={uiSchema}
        onSubmit={onSubmit}
      />
    );

    userEvent.click(screen.getByRole('button', { name: 'Submit' }));
    expect(onSubmit).toHaveBeenCalled();
  });
});
