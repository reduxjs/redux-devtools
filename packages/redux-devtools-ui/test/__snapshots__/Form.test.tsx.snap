// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<form
  class="sc-ftTHYK eHXaTW"
>
  <div
    class="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        Example form
      </legend>
      <p
        class="field-description"
        id="root__description"
      >
        A simple form example.
      </p>
      <div
        class="form-group field field-string"
      >
        <label
          class="control-label"
          for="root_name"
        >
          Full name
          <span
            class="required"
          >
            *
          </span>
        </label>
        <input
          class="form-control"
          id="root_name"
          label="Full name"
          placeholder=""
          required=""
          type="text"
          value="Chuck Norris"
        />
      </div>
      <div
        class="form-group field field-integer"
      >
        <label
          class="control-label"
          for="root_age"
        >
          Age
        </label>
        <input
          class="form-control"
          id="root_age"
          label="Age"
          placeholder=""
          type="number"
          value="75"
        />
      </div>
      <div
        class="form-group field field-string"
      >
        <label
          class="control-label"
          for="root_bio"
        >
          Bio
        </label>
        <textarea
          class="form-control"
          id="root_bio"
          placeholder=""
        >
          Roundhouse kicking asses since 1940
        </textarea>
      </div>
      <div
        class="form-group field field-string"
      >
        <label
          class="control-label"
          for="root_password"
        >
          Password
        </label>
        <input
          class="form-control"
          id="root_password"
          label="Password"
          placeholder=""
          type="password"
          value="noneed"
        />
        <p
          class="help-block"
          id="root_password__help"
        >
          Hint: Make it strong!
        </p>
      </div>
      <div
        class="form-group field field-array"
      >
        <label
          class="control-label"
          for="root_multipleChoicesList"
        >
          A multiple choices list
        </label>
        <div
          class="checkboxes"
          id="root_multipleChoicesList"
        >
          <div
            class="checkbox "
          >
            <label>
              <span>
                <input
                  id="root_multipleChoicesList_0"
                  type="checkbox"
                />
                <span>
                  foo
                </span>
              </span>
            </label>
          </div>
          <div
            class="checkbox "
          >
            <label>
              <span>
                <input
                  id="root_multipleChoicesList_1"
                  type="checkbox"
                />
                <span>
                  bar
                </span>
              </span>
            </label>
          </div>
          <div
            class="checkbox "
          >
            <label>
              <span>
                <input
                  id="root_multipleChoicesList_2"
                  type="checkbox"
                />
                <span>
                  fuzz
                </span>
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        class="form-group field field-number"
      >
        <label
          class="control-label"
          for="root_numberEnum"
        >
          Number enum
        </label>
        <div
          class=" css-18o3wl1-container"
          id="root_numberEnum"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class=" css-10g0wlc-control"
          >
            <div
              class=" css-1gpx584-ValueContainer"
            >
              <div
                class=" css-1m6ztbo-placeholder"
                id="react-select-2-placeholder"
              />
              <div
                class=" css-mh08on-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-2-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-required="false"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-2-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-19norqf-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-17vo4tq-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="form-group field field-number"
      >
        <label
          class="control-label"
          for="root_numberEnumRadio"
        >
          Number enum
        </label>
        <div
          class="field-radio-group"
          id="root_numberEnumRadio"
        >
          <label
            class="radio-inline "
          >
            <span>
              <input
                name="0.25546350798039463"
                type="radio"
                value="1"
              />
              <span>
                1
              </span>
            </span>
          </label>
          <label
            class="radio-inline "
          >
            <span>
              <input
                name="0.25546350798039463"
                type="radio"
                value="2"
              />
              <span>
                2
              </span>
            </span>
          </label>
          <label
            class="radio-inline "
          >
            <span>
              <input
                name="0.25546350798039463"
                type="radio"
                value="3"
              />
              <span>
                3
              </span>
            </span>
          </label>
        </div>
      </div>
      <div
        class="form-group field field-integer"
      >
        <label
          class="control-label"
          for="root_integerRange"
        >
          Integer range
        </label>
        <div
          class="sc-gKPRtg eacnbS"
        >
          <div
            class="sc-iBYQkv gUWPJm"
          >
            <input
              id="root_integerRange"
              max="100"
              min="42"
              placeholder=""
              type="range"
              uischema="[object Object]"
              value="52"
            />
            <div>
              52
            </div>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
  <div
    class="sc-eDvSVe cLxgTT"
  >
    <button
      class="sc-dkrFOg gBSOcI"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Form renders with no button 1`] = `
<form
  class="sc-ftTHYK eHXaTW"
>
  <div
    class="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        Example form
      </legend>
      <p
        class="field-description"
        id="root__description"
      >
        A simple form example.
      </p>
      <div
        class="form-group field field-string"
      >
        <label
          class="control-label"
          for="root_name"
        >
          Full name
          <span
            class="required"
          >
            *
          </span>
        </label>
        <input
          class="form-control"
          id="root_name"
          label="Full name"
          placeholder=""
          required=""
          type="text"
          value="Chuck Norris"
        />
      </div>
      <div
        class="form-group field field-integer"
      >
        <label
          class="control-label"
          for="root_age"
        >
          Age
        </label>
        <input
          class="form-control"
          id="root_age"
          label="Age"
          placeholder=""
          type="number"
          value="75"
        />
      </div>
      <div
        class="form-group field field-string"
      >
        <label
          class="control-label"
          for="root_bio"
        >
          Bio
        </label>
        <textarea
          class="form-control"
          id="root_bio"
          placeholder=""
        >
          Roundhouse kicking asses since 1940
        </textarea>
      </div>
      <div
        class="form-group field field-string"
      >
        <label
          class="control-label"
          for="root_password"
        >
          Password
        </label>
        <input
          class="form-control"
          id="root_password"
          label="Password"
          placeholder=""
          type="password"
          value="noneed"
        />
        <p
          class="help-block"
          id="root_password__help"
        >
          Hint: Make it strong!
        </p>
      </div>
      <div
        class="form-group field field-array"
      >
        <label
          class="control-label"
          for="root_multipleChoicesList"
        >
          A multiple choices list
        </label>
        <div
          class="checkboxes"
          id="root_multipleChoicesList"
        >
          <div
            class="checkbox "
          >
            <label>
              <span>
                <input
                  id="root_multipleChoicesList_0"
                  type="checkbox"
                />
                <span>
                  foo
                </span>
              </span>
            </label>
          </div>
          <div
            class="checkbox "
          >
            <label>
              <span>
                <input
                  id="root_multipleChoicesList_1"
                  type="checkbox"
                />
                <span>
                  bar
                </span>
              </span>
            </label>
          </div>
          <div
            class="checkbox "
          >
            <label>
              <span>
                <input
                  id="root_multipleChoicesList_2"
                  type="checkbox"
                />
                <span>
                  fuzz
                </span>
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        class="form-group field field-number"
      >
        <label
          class="control-label"
          for="root_numberEnum"
        >
          Number enum
        </label>
        <div
          class=" css-18o3wl1-container"
          id="root_numberEnum"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-4-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class=" css-10g0wlc-control"
          >
            <div
              class=" css-1gpx584-ValueContainer"
            >
              <div
                class=" css-1m6ztbo-placeholder"
                id="react-select-4-placeholder"
              />
              <div
                class=" css-mh08on-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-4-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-required="false"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-4-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-19norqf-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-17vo4tq-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="form-group field field-number"
      >
        <label
          class="control-label"
          for="root_numberEnumRadio"
        >
          Number enum
        </label>
        <div
          class="field-radio-group"
          id="root_numberEnumRadio"
        >
          <label
            class="radio-inline "
          >
            <span>
              <input
                name="0.25546350798039463"
                type="radio"
                value="1"
              />
              <span>
                1
              </span>
            </span>
          </label>
          <label
            class="radio-inline "
          >
            <span>
              <input
                name="0.25546350798039463"
                type="radio"
                value="2"
              />
              <span>
                2
              </span>
            </span>
          </label>
          <label
            class="radio-inline "
          >
            <span>
              <input
                name="0.25546350798039463"
                type="radio"
                value="3"
              />
              <span>
                3
              </span>
            </span>
          </label>
        </div>
      </div>
      <div
        class="form-group field field-integer"
      >
        <label
          class="control-label"
          for="root_integerRange"
        >
          Integer range
        </label>
        <div
          class="sc-gKPRtg eacnbS"
        >
          <div
            class="sc-iBYQkv gUWPJm"
          >
            <input
              id="root_integerRange"
              max="100"
              min="42"
              placeholder=""
              type="range"
              uischema="[object Object]"
              value="52"
            />
            <div>
              52
            </div>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
  <noscript />
</form>
`;

exports[`Form renders with primary button 1`] = `
<form
  class="sc-ftTHYK eHXaTW"
>
  <div
    class="form-group field field-object"
  >
    <fieldset
      id="root"
    >
      <legend
        id="root__title"
      >
        Example form
      </legend>
      <p
        class="field-description"
        id="root__description"
      >
        A simple form example.
      </p>
      <div
        class="form-group field field-string"
      >
        <label
          class="control-label"
          for="root_name"
        >
          Full name
          <span
            class="required"
          >
            *
          </span>
        </label>
        <input
          class="form-control"
          id="root_name"
          label="Full name"
          placeholder=""
          required=""
          type="text"
          value="Chuck Norris"
        />
      </div>
      <div
        class="form-group field field-integer"
      >
        <label
          class="control-label"
          for="root_age"
        >
          Age
        </label>
        <input
          class="form-control"
          id="root_age"
          label="Age"
          placeholder=""
          type="number"
          value="75"
        />
      </div>
      <div
        class="form-group field field-string"
      >
        <label
          class="control-label"
          for="root_bio"
        >
          Bio
        </label>
        <textarea
          class="form-control"
          id="root_bio"
          placeholder=""
        >
          Roundhouse kicking asses since 1940
        </textarea>
      </div>
      <div
        class="form-group field field-string"
      >
        <label
          class="control-label"
          for="root_password"
        >
          Password
        </label>
        <input
          class="form-control"
          id="root_password"
          label="Password"
          placeholder=""
          type="password"
          value="noneed"
        />
        <p
          class="help-block"
          id="root_password__help"
        >
          Hint: Make it strong!
        </p>
      </div>
      <div
        class="form-group field field-array"
      >
        <label
          class="control-label"
          for="root_multipleChoicesList"
        >
          A multiple choices list
        </label>
        <div
          class="checkboxes"
          id="root_multipleChoicesList"
        >
          <div
            class="checkbox "
          >
            <label>
              <span>
                <input
                  id="root_multipleChoicesList_0"
                  type="checkbox"
                />
                <span>
                  foo
                </span>
              </span>
            </label>
          </div>
          <div
            class="checkbox "
          >
            <label>
              <span>
                <input
                  id="root_multipleChoicesList_1"
                  type="checkbox"
                />
                <span>
                  bar
                </span>
              </span>
            </label>
          </div>
          <div
            class="checkbox "
          >
            <label>
              <span>
                <input
                  id="root_multipleChoicesList_2"
                  type="checkbox"
                />
                <span>
                  fuzz
                </span>
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        class="form-group field field-number"
      >
        <label
          class="control-label"
          for="root_numberEnum"
        >
          Number enum
        </label>
        <div
          class=" css-18o3wl1-container"
          id="root_numberEnum"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-3-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class=" css-10g0wlc-control"
          >
            <div
              class=" css-1gpx584-ValueContainer"
            >
              <div
                class=" css-1m6ztbo-placeholder"
                id="react-select-3-placeholder"
              />
              <div
                class=" css-mh08on-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-3-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-required="false"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-3-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-19norqf-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-17vo4tq-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="form-group field field-number"
      >
        <label
          class="control-label"
          for="root_numberEnumRadio"
        >
          Number enum
        </label>
        <div
          class="field-radio-group"
          id="root_numberEnumRadio"
        >
          <label
            class="radio-inline "
          >
            <span>
              <input
                name="0.25546350798039463"
                type="radio"
                value="1"
              />
              <span>
                1
              </span>
            </span>
          </label>
          <label
            class="radio-inline "
          >
            <span>
              <input
                name="0.25546350798039463"
                type="radio"
                value="2"
              />
              <span>
                2
              </span>
            </span>
          </label>
          <label
            class="radio-inline "
          >
            <span>
              <input
                name="0.25546350798039463"
                type="radio"
                value="3"
              />
              <span>
                3
              </span>
            </span>
          </label>
        </div>
      </div>
      <div
        class="form-group field field-integer"
      >
        <label
          class="control-label"
          for="root_integerRange"
        >
          Integer range
        </label>
        <div
          class="sc-gKPRtg eacnbS"
        >
          <div
            class="sc-iBYQkv gUWPJm"
          >
            <input
              id="root_integerRange"
              max="100"
              min="42"
              placeholder=""
              type="range"
              uischema="[object Object]"
              value="52"
            />
            <div>
              52
            </div>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
  <div
    class="sc-eDvSVe cLxgTT"
  >
    <button
      class="sc-dkrFOg bUbFU"
      type="submit"
    >
      Custom button
    </button>
  </div>
</form>
`;
